src/commands/distribute/lib/mc-fus-uploader/mc-fus-uploader.ts

```diff
@@ -0,0 +1,777 @@
+    this.path = path;
+  }
+
+  get size(): number {
+    const stats = fs.statSync(this.path);
+    return stats["size"];
+  }
+
+  get name(): string {
+    return Path.basename(this.path);
+  }
+
+  slice(start: number, end: number): Buffer {
+    const data = Buffer.alloc(end - start);
+    const fd = fs.openSync(this.path, "r");
+    fs.readSync(fd, data, 0, data.length, start);
+    return data;
+  }
+}
+
+class HttpError extends Error {
+  readonly status: number;
+  readonly statusText: string;
+
+  constructor(status: number, statusText: string) {
+    super(statusText);
+    Object.setPrototypeOf(this, new.target.prototype);
+    this.status = status;
+    this.statusText = statusText;
+  }
+}
+
+export class McFusNodeUploader implements McFusUploader {
+  ambiguousProgress: number = 0;
+  progressUpdateRate: number = 0;
+  maxNumberOfConcurrentUploads: number = 10;
+
+  readonly uploadBaseUrls: any = {
+    CancelUpload: "upload/cancel/",
+    SetMetadata: "upload/set_metadata/",
+    UploadChunk: "upload/upload_chunk/",
+    UploadFinished: "upload/finished/",
+    UploadStatus: "upload/status/",
+  };
+
+  readonly uploadData: IUploadData = {
+    AssetId: "00000000-0000-0000-0000-000000000000",
+    BlobPartitions: 0,
+    CallbackUrl: "",
+    CorrelationId: "00000000-0000-0000-0000-000000000000",
+    CorrelationVector: "",
+    ChunkSize: 0,
+    LogToConsole: false,
+    Tenant: "",
+    UrlEncodedToken: "",
+    TotalBlocks: 0,
+    UploadDomain: "",
+  };
+
+  readonly uploadStatus: IUploadStatus = {
+    AutoRetryCount: 0,
+    AverageSpeed: 0,
+    BlocksCompleted: 0,
+    ChunksFailedCount: 0,
+    ChunkQueue: [],
+    Connected: true,
+    EndTime: new Date(),
+    InflightSet: new Set(),
+    AbortController: new AbortController(),
+    MaxErrorCount: 20,
+    ServiceCallback: {
+      AutoRetryCount: 5,
+      AutoRetryDelay: 1,
+      FailureCount: 0,
+    },
+    StartTime: new Date(),
+    State: McFusUploadState.New,
+    TransferQueueRate: [],
+  };
+
+  readonly eventHandlers: any = {
+    onProgressChanged: (progress: IProgress) => {},
+    onCompleted: (uploadStats: IUploadStats) => {},
+    onResumeRestart: (onResumeStartParams: IOnResumeStartParams) => {},
+    onMessage: (message: string, properties: LogProperties, messageLevel: McFusMessageLevel) => {},
+    onStateChanged: (state: McFusUploadState) => {},
+  };
+
+  constructor(args: IInitializeSettings) {
+    this.initializeUpload(args);
+  }
+
+  calculateAverageSpeed() {
+    if (this.uploadStatus.TransferQueueRate.length === 0) {
+      return 0;
+    }
+
+    let rateSum = 0;
+
+    for (const transferQueueRate of this.uploadStatus.TransferQueueRate) {
+      rateSum += transferQueueRate;
+    }
+
+    return rateSum / this.uploadStatus.TransferQueueRate.length;
+  }
+
+  calculateRate() {
+    // Get the elapsed time in seconds
+    const diff = new Date().getTime() - this.uploadStatus.StartTime.getTime();
+    const seconds = diff / 1000;
```

Comment:
Let's return 0 if `seconds` is 0 by accident.

Reply:
Fixed in https://github.com/microsoft/appcenter-cli/pull/929